ICE when expanding `matches!` macro in `core` 1.49

```rust
macro_rules! matches {
    ($expression:expr, $( $pattern:pat )|+ $( if $guard: expr )? $(,)?) => {
        match $expression {
            $( $pattern )|+ $( if $guard )? => true,
            _ => false
        }
    }
}

#[stable(feature = "ascii_ctype_on_intrinsics", since = "1.24.0")]
#[rustc_const_stable(feature = "const_ascii_ctype_on_intrinsics", since = "1.47.0")]
#[inline]
pub const fn is_ascii_alphabetic(c: char) -> bool {
    matches!(c, b'A'..=b'Z' | b'a'..=b'z')
}
```

```
crab1: internal compiler error: Segmentation fault
0x3b65d20 crash_signal
	../../gcc/toplev.cc:314
0x2019d74 Rust::MatchedFragment::operator=(Rust::MatchedFragment const&)
	../../gcc/rust/expand/rust-macro-expand.h:53
0x2019277 Rust::SubstituteCtx::substitute_repetition(unsigned long, unsigned long, std::unique_ptr<Rust::AST::Token, std::default_delete<Rust::AST::Token> >)
	../../gcc/rust/expand/rust-macro-substitute-ctx.cc:155
0x20198c4 Rust::SubstituteCtx::substitute_token(unsigned long)
	../../gcc/rust/expand/rust-macro-substitute-ctx.cc:261
0x2019a8d Rust::SubstituteCtx::substitute_tokens()
	../../gcc/rust/expand/rust-macro-substitute-ctx.cc:294
0x1fa70bf Rust::MacroExpander::transcribe_rule(Rust::AST::MacroRule&, Rust::AST::DelimTokenTree&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, Rust::MatchedFragmentContainer, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, Rust::MatchedFragmentContainer> > >&, bool, Rust::MacroExpander::ContextType)
	../../gcc/rust/expand/rust-macro-expand.cc:980
0x1fa3d1f Rust::MacroExpander::expand_decl_macro(Location, Rust::AST::MacroInvocData&, Rust::AST::MacroRulesDefinition&, bool)
	../../gcc/rust/expand/rust-macro-expand.cc:111
0x1fa473c Rust::MacroExpander::expand_invoc(Rust::AST::MacroInvocation&, bool)
	../../gcc/rust/expand/rust-macro-expand.cc:283
0x200810e Rust::ExpandVisitor::visit(Rust::AST::MacroInvocation&)
	../../gcc/rust/expand/rust-expand-visitor.cc:256
0x1e72dab Rust::AST::MacroInvocation::accept_vis(Rust::AST::ASTVisitor&)
	../../gcc/rust/ast/rust-ast.cc:4793
0x2007672 Rust::ExpandVisitor::maybe_expand_expr(std::unique_ptr<Rust::AST::Expr, std::default_delete<Rust::AST::Expr> >&)
	../../gcc/rust/expand/rust-expand-visitor.cc:62
0x2009150 Rust::ExpandVisitor::visit(Rust::AST::BlockExpr&)
	../../gcc/rust/expand/rust-expand-visitor.cc:551
0x1e72157 Rust::AST::BlockExpr::accept_vis(Rust::AST::ASTVisitor&)
	../../gcc/rust/ast/rust-ast.cc:4433
0x200e95d void Rust::ExpandVisitor::visit<Rust::AST::BlockExpr>(std::unique_ptr<Rust::AST::BlockExpr, std::default_delete<Rust::AST::BlockExpr> >&)
	../../gcc/rust/expand/rust-expand-visitor.h:136
0x200a05d Rust::ExpandVisitor::visit(Rust::AST::Function&)
	../../gcc/rust/expand/rust-expand-visitor.cc:832
0x1e72811 Rust::AST::Function::accept_vis(Rust::AST::ASTVisitor&)
	../../gcc/rust/ast/rust-ast.cc:4631
0x200748a Rust::ExpandVisitor::go(Rust::AST::Crate&)
	../../gcc/rust/expand/rust-expand-visitor.cc:36
0x1f80d96 Rust::Session::expansion(Rust::AST::Crate&)
	../../gcc/rust/rust-session-manager.cc:863
0x1f7ffbb Rust::Session::compile_crate(char const*)
	../../gcc/rust/rust-session-manager.cc:569
0x1f7f5d2 Rust::Session::handle_input_files(int, char const**)
	../../gcc/rust/rust-session-manager.cc:381
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
```

No longer produces an ICE, now produces
```
test.rs:14:17: sorry, unimplemented: type checking alternate patterns not supported
   14 |     matches!(c, b'A'..=b'Z' | b'a'..=b'z')
      |                 ^~~
test.rs:3:9: error: failed to type resolve expression
    3 |         match $expression {
      |         ^~~~~
```

--- a/gcc/testsuite/rust/execute/torture/matches_macro.rs
+++ b/gcc/testsuite/rust/execute/torture/matches_macro.rs
@@ -1,0 +1,30 @@
+macro_rules! matches {
+    ($expression:expr, $($pattern:pat)|+ $( if $guard:expr ),*) => {
+        match $expression {
+            $($pattern)|+ => true,
+            _ => false,
+        }
+    }
+}
+
+pub fn should_match() -> bool {
+    matches!(1, 1)
+}
+
+pub fn shouldnt() -> bool {
+    matches!(1, 2)
+}
+
+fn main() -> i32 {
+    let mut retval = 2;
+
+    if should_match() {
+        retval -= 1;
+    }
+
+    if !shouldnt() {
+        retval -= 1;
+    }
+
+    retval
+}

